cmd/gateway-gcs.go

```diff
@@ -268,17 +279,67 @@ func newGCSGateway(projectID string) (GatewayLayer, error) {
 		return nil, err
 	}
 
-	return &gcsGateway{
-		client:     client,
-		projectID:  projectID,
-		ctx:        ctx,
-		anonClient: anonClient,
-	}, nil
+	gateway := &gcsGateway{
+		client:       client,
+		projectID:    projectID,
+		ctx:          ctx,
+		anonClient:   anonClient,
+		cleanupEndCh: make(chan struct{}),
+	}
+	// Start background process to cleanup old files in minio.sys.temp
+	go gateway.CleanupGCSMinioPath()
+	return gateway, nil
+}
+
+// Cleanup old files in minio.sys.temp of the given bucket.
+func (l *gcsGateway) CleanupGCSMinioPathBucket(bucket string) {
+	it := l.client.Bucket(bucket).Objects(l.ctx, &storage.Query{Prefix: gcsMinioTempPath, Versions: false})
+	for {
+		attrs, err := it.Next()
+		if err != nil {
+			if err != iterator.Done {
+				errorIf(err, "Object listing error on bucket %s", bucket)
+			}
+			return
+		}
+		if time.Since(attrs.Updated) > gcsMultipartExpiry {
+			// Delete files older than 2 weeks.
+			err := l.client.Bucket(bucket).Object(attrs.Name).Delete(l.ctx)
+			if err != nil {
+				errorIf(err, "Unable to delete the object %s", attrs.Name)
+				return
+			}
+		}
+	}
+}
+
+// Cleanup old files in minio.sys.temp of all buckets.
+func (l *gcsGateway) CleanupGCSMinioPath() {
+	for {
+		it := l.client.Buckets(l.ctx, l.projectID)
+		for {
+			attrs, err := it.Next()
+			if err != nil {
+				if err != iterator.Done {
+					errorIf(err, "Bucket listing error")
+				}
+				break
+			}
+			l.CleanupGCSMinioPathBucket(attrs.Name)
+		}
+		select {
+		case <-l.cleanupEndCh:
```

Comment:
We can let the cleanup complete. Cleanup will be a quick process as the number of entries in minio.sys.temp will be less.